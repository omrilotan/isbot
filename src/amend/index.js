/**
 * Mutate given list of patter strings
 * @param {string[]} list
 * @returns {string[]}
 */
export function amend (list) {
  try {
    // Risk: Uses lookbehind assertion, avoid breakage in parsing by using RegExp constructor
    new RegExp('(?<! cu)bot').test('dangerbot') // eslint-disable-line prefer-regex-literals
  } catch (error) {
    // Skip regex fixes
    return list
  }

  [
    // Addresses: Cubot device
    ['bot', '(?<! cu)bot'],
    // Addresses: Android webview
    ['google', '(?<! (?:channel/|google/))google(?!(app|/google| pixel))'],
    // Addresses: libhttp browser
    ['http', '(?<!(?:lib))http'],
    // Addresses: java based browsers
    ['java', 'java(?!;)'],
    // Addresses: Yandex Search App
    ['search', '(?<! ya(?:yandex)?)search']
  ].forEach(
    ([search, replace]) => {
      const index = list.lastIndexOf(search)
      if (~index) {
        list.splice(index, 1, replace)
      }
    }
  )

  return list
}
