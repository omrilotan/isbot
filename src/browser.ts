import { isbot } from ".";

export default isbot;
