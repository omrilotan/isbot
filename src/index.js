import { Isbot } from './isbot/index.js'

const isbot = new Isbot()

export default isbot
