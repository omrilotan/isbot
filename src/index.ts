import patternsList from "./patterns.json";
import { fullPattern } from "./pattern";

/**
 * Naive bot pattern.
 */
const naivePattern = /bot|crawl|http|lighthouse|scan|search|spider/i;

let pattern: RegExp;
export function getPattern(): RegExp {
	if (pattern instanceof RegExp) {
		return pattern;
	}
	try {
		// Build this RegExp dynamically to avoid syntax errors in older engines.
		pattern = new RegExp(fullPattern, "i");
	} catch (error) {
		pattern = naivePattern;
	}
	return pattern;
}

/**
 * A list of bot identifiers to be used in a regular expression against user agent strings.
 */
export const list: string[] = patternsList;

/**
 * Check if the given user agent includes a bot pattern. Naive implementation (less accurate).
 */
export const isbotNaive = (userAgent?: string | null): boolean =>
	Boolean(userAgent) && naivePattern.test(userAgent);

/**
 * Check if the given user agent includes a bot pattern.
 */
export function isbot(userAgent?: string | null): boolean {
	return Boolean(userAgent) && getPattern().test(userAgent);
}

/**
 * Create a custom isbot function with a custom pattern.
 */
export const createIsbot =
	(customPattern: RegExp): ((userAgent?: string | null) => boolean) =>
	(userAgent: string): boolean =>
		Boolean(userAgent) && customPattern.test(userAgent);

/**
 * Create a custom isbot function with a custom pattern.
 */
export const createIsbotFromList = (
	list: string[],
): ((userAgent: string) => boolean) => {
	const pattern = new RegExp(list.join("|"), "i");
	return (userAgent: string): boolean =>
		Boolean(userAgent) && pattern.test(userAgent);
};

/**
 * Find the first part of the user agent that matches a bot pattern.
 */
export const isbotMatch = (userAgent?: string | null): string | null =>
	userAgent?.match(getPattern())?.[0] ?? null;

/**
 * Find all parts of the user agent that match a bot pattern.
 */
export const isbotMatches = (userAgent?: string | null): string[] =>
	list
		.map((part) => userAgent?.match(new RegExp(part, "i"))?.[0])
		.filter(Boolean);

/**
 * Find the first bot pattern that match the given user agent.
 */
export const isbotPattern = (userAgent?: string | null): string | null =>
	userAgent
		? (list.find((pattern) => new RegExp(pattern, "i").test(userAgent)) ?? null)
		: null;

/**
 * Find all bot patterns that match the given user agent.
 */
export const isbotPatterns = (userAgent?: string | null): string[] =>
	userAgent
		? list.filter((pattern) => new RegExp(pattern, "i").test(userAgent))
		: [];
