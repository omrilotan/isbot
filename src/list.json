[
  " daum[/|\\s]",
  " deusu/",
  "(?:^|\\s)site",
  "@[a-z]",
  "\\(at\\)[a-z]",
  "\\(github\\.com/",
  "\\[at\\][a-z]",
  "^12345",
  "^<",
  "^ace explorer",
  "^acoon",
  "^active",
  "^ad muncher",
  "^ahc/",
  "^anglesharp/",
  "^anonymous",
  "^apache",
  "^applicationhealthservice",
  "^arachni/",
  "^astute srm",
  "^avsdevicesdk/",
  "^axios/",
  "^bidtellect/",
  "^biglotron",
  "^bitdiscovery",
  "^blackboard safeassign",
  "^blocknote.net",
  "^blogtrottr/",
  "^braze sender",
  "^browsershots",
  "^cakephp",
  "^camo asset proxy",
  "^captivenetworksupport",
  "^castro",
  "^clamav[\\s/]",
  "^cobweb/",
  "^coccoc",
  "^coldfusion",
  "^collectd/",
  "^custom$",
  "^dap ",
  "^datadog agent/",
  "^davclnt",
  "^ddg_android/",
  "^deluge",
  "^discourse",
  "^dispatch/\\d",
  "^disqus/",
  "^duckduckgo",
  "^enigma browser",
  "^evernote clip resolver",
  "^facebook",
  "^faraday",
  "^fdm[/\\s]\\d",
  "^flashget",
  "^friendica",
  "^getright/",
  "^gigablastopensource",
  "^gobuster/",
  "^googal",
  "^goose",
  "^gozilla/",
  "^greenbrowser",
  "^hatena",
  "^hexometer",
  "^hobbit",
  "^hotzonu",
  "^hwcdn/",
  "^infox-wisg",
  "^ingrid/\\d",
  "^integrity/",
  "^invision",
  "^jeode/",
  "^jetbrains",
  "^jetty/",
  "^jigsaw",
  "^libwww",
  "^linkdex",
  "^lwp-",
  "^lwp::simple",
  "^mailchimp\\.com$",
  "^metauri",
  "^microsoft bits",
  "^microsoft data",
  "^microsoft office existence",
  "^microsoft office protocol discovery",
  "^microsoft windows network diagnostics",
  "^microsoft-cryptoapi",
  "^microsoft-webdav-miniredir",
  "^monit",
  "^movabletype",
  "^mozilla/\\d\\.\\d \\(compatible;?\\)$",
  "^my browser$",
  "^navermailapp",
  "^netsurf",
  "^nibbler",
  "^ning",
  "^node-superagent",
  "^notetextview",
  "^nuzzel",
  "^octopus",
  "^offline explorer",
  "^ossproxy",
  "^pagething",
  "^panscient",
  "^pearltrees",
  "^perimeterx",
  "^photon/",
  "^php",
  "^postman",
  "^postrank",
  "^prometheus/",
  "^python",
  "^ramblermail",
  "^restsharp/",
  "^robozilla/",
  "^ruby$",
  "^scrapy",
  "^seo",
  "^shareaza",
  "^shockwave flash",
  "^shortlinktranslate",
  "^sistrix",
  "^sixy.ch/",
  "^smallproxy",
  "^snap$",
  "^snapchat",
  "^space bison",
  "^spotify/",
  "^sprinklr",
  "^svn",
  "^swcd ",
  "^t-online browser",
  "^taringa",
  "^test certificate info",
  "^the knowledge ai",
  "^thinklab",
  "^thumb",
  "^traackr.com",
  "^transmission",
  "^tumblr/",
  "^ucmore",
  "^upflow/",
  "^user_agent",
  "^vbulletin",
  "^venus/fedoraplanet",
  "^viber$",
  "^w3c",
  "^webbandit/",
  "^webcollage/",
  "^webcopier",
  "^wget",
  "^whatsapp",
  "^whatweb",
  "^www-mechanize",
  "^xenu link sleuth",
  "^xymon",
  "^yahoo",
  "^yandex",
  "^zabbix",
  "^zdm/\\d",
  "^zeushdthree",
  "^zmeu$",
  "adbeat\\.com",
  "amiga",
  "appinsights",
  "archive",
  "ask jeeves/teoma",
  "bit.ly/",
  "bluecoat drtr",
  "bot",
  "browsex",
  "burpcollaborator",
  "capture",
  "catch",
  "check",
  "chrome-lighthouse",
  "chromeframe",
  "client",
  "cloud",
  "crawl",
  "cron",
  "daemon",
  "dareboost",
  "datanyze",
  "dataprovider",
  "dejaclick",
  "dmbrowser",
  "download",
  "email",
  "evc-batch/",
  "feed",
  "fetch",
  "finder",
  "firephp",
  "freesafeip",
  "ghost",
  "gomezagent",
  "google",
  "headlesschrome/",
  "http",
  "httrack",
  "hubspot marketing grader",
  "hydra",
  "ibisbrowser",
  "images",
  "index",
  "ips-agent",
  "java",
  "jorgee",
  "library",
  "mail\\.ru/",
  "manager",
  "monitor",
  "neustar wpm",
  "news",
  "nutch",
  "offbyone",
  "optimize",
  "pagespeed",
  "parse",
  "perl",
  "phantom",
  "pingdom",
  "powermarks",
  "preview",
  "probe",
  "ptst[/ ]\\d",
  "reader",
  "rigor",
  "rss",
  "scan",
  "scrape",
  "search",
  "server",
  "sogou",
  "sparkler/",
  "spider",
  "statuscake",
  "stumbleupon\\.com",
  "supercleaner",
  "synapse",
  "synthetic",
  "toolbar",
  "torrent",
  "tracemyfile",
  "transcoder",
  "trendsmapresolver",
  "twingly recon",
  "url",
  "valid",
  "virtuoso",
  "wappalyzer",
  "webglance",
  "webkit2png",
  "websitemetadataretriever",
  "whatcms/",
  "wordpress",
  "zgrab"
]
