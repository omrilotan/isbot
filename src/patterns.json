[
  " daum[ /]",
  " deusu/",
  " yadirectfetcher",
  "(?:^|[^g])news(?!sapphire)",
  "(?<! (?:channel/|google/))google(?!(app|/google| pixel))",
  "(?<! cu)bots?(?:\\b|_)",
  "(?<!(?:lib))http",
  "(?<![hg]m)score",
  "@[a-z][\\w-]+\\.",
  "\\(\\)",
  "\\.com\\b",
  "\\btime/",
  "^<",
  "^[\\w \\.\\-\\(?:\\):]+(?:/v?\\d+(?:\\.\\d+)?(?:\\.\\d{1,10})*?)?(?:,|$)",
  "^[^ ]{50,}$",
  "^\\d+\\b",
  "^\\w*search\\b",
  "^\\w+/[\\w\\(\\)]*$",
  "^active",
  "^ad muncher",
  "^amaya",
  "^avsdevicesdk/",
  "^biglotron",
  "^bot",
  "^bw/",
  "^clamav[ /]",
  "^client/",
  "^cobweb/",
  "^custom",
  "^ddg[_-]android",
  "^discourse",
  "^dispatch/\\d",
  "^downcast/",
  "^duckduckgo",
  "^facebook",
  "^getright/",
  "^gozilla/",
  "^hobbit",
  "^hotzonu",
  "^hwcdn/",
  "^jeode/",
  "^jetty/",
  "^jigsaw",
  "^microsoft bits",
  "^movabletype",
  "^mozilla/5\\.0\\s[a-z\\.-]+$",
  "^mozilla/\\d\\.\\d \\(compatible;?\\)$",
  "^mozilla/\\d\\.\\d \\w*$",
  "^navermailapp",
  "^netsurf",
  "^offline",
  "^owler",
  "^php",
  "^postman",
  "^python",
  "^rank",
  "^read",
  "^reed",
  "^rest",
  "^rss",
  "^snapchat",
  "^space bison",
  "^svn",
  "^swcd ",
  "^taringa",
  "^thumbor/",
  "^track",
  "^valid",
  "^w3c",
  "^webbandit/",
  "^webcopier",
  "^wget",
  "^whatsapp",
  "^wordpress",
  "^xenu link sleuth",
  "^yahoo",
  "^yandex",
  "^zdm/\\d",
  "^zoom marketplace/",
  "^{{.*}}$",
  "adscanner/",
  "analyzer",
  "archive",
  "ask jeeves/teoma",
  "bit\\.ly/",
  "bluecoat drtr",
  "browsex",
  "burpcollaborator",
  "capture",
  "catch",
  "check\\b",
  "checker",
  "chrome-lighthouse",
  "chromeframe",
  "classifier",
  "cloudflare",
  "convertify",
  "crawl",
  "cypress/",
  "dareboost",
  "datanyze",
  "dejaclick",
  "detect",
  "dmbrowser",
  "download",
  "evc-batch/",
  "exaleadcloudview",
  "feed",
  "firephp",
  "functionize",
  "gomezagent",
  "headless",
  "httrack",
  "hubspot marketing grader",
  "hydra",
  "ibisbrowser",
  "images",
  "infrawatch",
  "insight",
  "inspect",
  "iplabel",
  "ips-agent",
  "java(?!;)",
  "jsjcw_scanner",
  "library",
  "linkcheck",
  "mail\\.ru/",
  "manager",
  "measure",
  "neustar wpm",
  "node",
  "nutch",
  "offbyone",
  "optimize",
  "pageburst",
  "pagespeed",
  "parser",
  "perl",
  "phantomjs",
  "pingdom",
  "powermarks",
  "preview",
  "proxy",
  "ptst[ /]\\d",
  "reputation",
  "resolver",
  "retriever",
  "rexx;",
  "rigor",
  "rss\\b",
  "scanner\\.",
  "scrape",
  "server",
  "sogou",
  "sparkler/",
  "speedcurve",
  "spider",
  "splash",
  "statuscake",
  "supercleaner",
  "synapse",
  "synthetic",
  "tools",
  "torrent",
  "trace",
  "transcoder",
  "url",
  "virtuoso",
  "wappalyzer",
  "webglance",
  "webkit2png",
  "whatcms/",
  "zgrab"
]
