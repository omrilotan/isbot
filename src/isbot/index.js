import list from '../list.json' assert { type: 'json' }
import { amend } from '../amend/index.js'

amend(list)

const flags = 'i'

/**
 * Test user agents for matching patterns
 */
export class Isbot {
  /**
   * @type {string[]}
   */
  #list

  /**
   * @type {RegExp}
   */
  #pattern

  constructor (patterns) {
    this.#list = patterns || list.slice()
    this.#update()

    const isbot = ua => this.test(ua)

    return Object.defineProperties(
      isbot,
      Object.entries(Object.getOwnPropertyDescriptors(Isbot.prototype)).reduce(
        (accumulator, [prop, descriptor]) => {
          if (typeof descriptor.value === 'function') {
            Object.assign(
              accumulator,
              { [prop]: { value: this[prop].bind(this) } }
            )
          }
          if (typeof descriptor.get === 'function') {
            Object.assign(
              accumulator,
              { [prop]: { get: () => this[prop] } }
            )
          }
          return accumulator
        },
        {}
      )
    )
  }

  /**
   * Recreate the pattern from rules list
   */
  #update () {
    this.#pattern = new RegExp(
      this.#list.join('|'),
      flags
    )
  }

  /**
   * Find the first index of an existing rule or -1 if not found
   * @param  {string} rule
   * @returns {number}
   */
  #index (rule) {
    return this.#list.indexOf(rule.toLowerCase())
  }

  /**
   * Get a clone of the pattern
   * @type RegExp
   */
  get pattern () {
    return new RegExp(this.#pattern)
  }

  /**
   * Match given string against out pattern
   * @param  {string} ua User Agent string
   * @returns {boolean}
   */
  test (ua) {
    return Boolean(ua) && this.#pattern.test(ua)
  }

  /**
   * Get the match for strings' known crawler pattern
   * @param  {string} ua User Agent string
   * @returns {string|null}
   */
  find (ua = '') {
    const match = ua.match(this.#pattern)
    return match && match[0]
  }

  /**
   * Get the patterns that match user agent string if any
   * @param  {string} ua User Agent string
   * @returns {string[]}
   */
  matches (ua = '') {
    return this.#list.filter(
      entry => new RegExp(entry, flags).test(ua)
    )
  }

  /**
   * Clear all patterns that match user agent
   * @param  {string} ua User Agent string
   * @returns {void}
   */
  clear (ua = '') {
    this.exclude(this.matches(ua))
  }

  /**
   * Extent patterns for known crawlers
   * @param  {string[]} filters
   * @returns {void}
   */
  extend (filters = []) {
    [].push.apply(
      this.#list,
      filters.filter(
        rule => this.#index(rule) === -1
      ).map(
        filter => filter.toLowerCase()
      )
    )
    this.#update()
  }

  /**
   * Exclude patterns from bot pattern rule
   * @param  {string[]} filters
   * @returns {void}
   */
  exclude (filters = []) {
    let { length } = filters
    while (length--) {
      const index = this.#index(filters[length])
      if (index > -1) {
        this.#list.splice(index, 1)
      }
    }
    this.#update()
  }

  /**
   * Create a new Isbot instance using given list or self's list
   * @param  {string[]} [list]
   * @returns {Isbot}
   */
  spawn (list) {
    return new Isbot(list || this.#list)
  }
}
