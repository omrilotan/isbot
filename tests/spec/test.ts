import {
	getPattern,
	list,
	isbot,
	isbotNaive,
	isbotMatch,
	isbotMatches,
	isbotPattern,
	isbotPatterns,
	createIsbot,
	createIsbotFromList,
} from "../../src";
import { fullPattern, regularExpression } from "../../src/pattern";
import { crawlers, browsers } from "../../fixtures";
let isbotInstance: any;

const BOT_USER_AGENT_EXAMPLE =
	"Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)";
const BROWSER_USER_AGENT_EXAMPLE =
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91 Safari/537.36";

const USER_AGENT_COMMON = [
	"Ada Chat Bot/1.0 Request Block",
	"Mozilla/5.0 (compatible; Yahoo! Slurp; http://help.yahoo.com/help/us/ysearch/slurp)",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4590.2 Safari/537.36 Chrome-Lighthouse",
];
const USER_AGENT_GOTCHAS = [
	"Mozilla/5.0 (Linux; Android 10; CUBOT_X30) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.85 Mobile Safari/537.36",
	"PS4Application libhttp/1.000 (PS4) CoreMedia libhttp/6.72 (PlayStation 4)",
];

describe("isbot", () => {
	describe("features", () => {
		test("pattern: pattern is a regex", () => {
			expect(getPattern()).toBeInstanceOf(RegExp);
		});
		test("list: list is an array", () => {
			expect(list).toBeInstanceOf(Array);
			expect(list.every((item) => typeof item === "string")).toBe(true);
		});
		test("isbot: bot user agect string is recognised as bot", () => {
			expect(isbot(BOT_USER_AGENT_EXAMPLE)).toBe(true);
		});
		test("isbotMatch: find pattern in bot user agent string", () => {
			expect(isbotMatch(BOT_USER_AGENT_EXAMPLE)).toBe("Google");
		});
		test("isbotMatches: find all patterns in bot user agent string", () => {
			expect(isbotMatches(BOT_USER_AGENT_EXAMPLE)).toContain("Google");
			expect(isbotMatches(BOT_USER_AGENT_EXAMPLE)).toHaveLength(3);
		});
		test("isbotPattern: find first pattern in bot user agent string", () => {
			expect(isbotPattern(BOT_USER_AGENT_EXAMPLE)).toBe(
				"(?<! (?:channel/|google/))google(?!(app|/google| pixel))",
			);
		});
		test("isbotPatterns: find all patterns in bot user agent string", () => {
			expect(isbotPatterns(BOT_USER_AGENT_EXAMPLE)).toContain(
				"(?<! (?:channel/|google/))google(?!(app|/google| pixel))",
			);
			expect(isbotPatterns(BOT_USER_AGENT_EXAMPLE)).toHaveLength(3);
		});
		test("createIsbot: create custom isbot function with custom pattern", () => {
			const customIsbot = createIsbot(/bot/i);
			expect(customIsbot(BOT_USER_AGENT_EXAMPLE)).toBe(true);
		});
		test("createIsbotFromList: create custom isbot function with custom pattern", () => {
			const ChromeLighthouseUserAgentStrings: string[] = [
				"mozilla/5.0 (macintosh; intel mac os x 10_15_7) applewebkit/537.36 (khtml, like gecko) chrome/94.0.4590.2 safari/537.36 chrome-lighthouse",
				"mozilla/5.0 (linux; android 7.0; moto g (4)) applewebkit/537.36 (khtml, like gecko) chrome/94.0.4590.2 mobile safari/537.36 chrome-lighthouse",
			];
			const patternsToRemove: Set<string> = new Set(
				ChromeLighthouseUserAgentStrings.map(isbotMatches).flat(),
			);
			const isbot2 = createIsbotFromList(
				list.filter(
					(record: string): boolean => patternsToRemove.has(record) === false,
				),
			);
			const [ua] = ChromeLighthouseUserAgentStrings;
			expect(isbot(ua)).toBe(true);
			expect(isbot2(ua)).toBe(false);
		});
		test.each([null, undefined, ""])(
			"all functions can accept %p",
			(value: string | null | undefined) => {
				expect(isbot(value)).toBe(false);
				expect(isbotMatch(value)).toBe(null);
				expect(isbotMatches(value)).toEqual([]);
				expect(isbotPattern(value)).toBe(null);
				expect(isbotPatterns(value)).toEqual([]);
			},
		);
	});

	describe("isbotNaive", () => {
		test.each([75])(
			"a large number of user agent strings can be detected (>%s%)",
			(percent) => {
				const ratio =
					crawlers.filter((ua) => isbotNaive(ua)).length / crawlers.length;
				expect(ratio).toBeLessThan(1);
				expect(ratio).toBeGreaterThan(percent / 100);
			},
		);
		test.each([1])(
			"a small number of browsers is falsly detected as bots (<%s%)",
			(percent) => {
				const ratio =
					browsers.filter((ua) => isbotNaive(ua)).length / browsers.length;
				expect(ratio).toBeGreaterThan(0);
				expect(ratio).toBeLessThan(percent / 100);
			},
		);
	});

	describe("regex fallback", () => {
		beforeAll(async () => {
			jest
				.spyOn(globalThis, "RegExp")
				.mockImplementation((pattern, flags): RegExp => {
					if ((pattern as string).includes?.("?<!")) {
						throw new Error("Invalid regex");
					}
					return new RegExp(pattern, flags);
				});
			const mdl = await import("../../index.js");
			if (!mdl) {
				throw new Error("Module not found");
			}
			isbotInstance = mdl.isbot as ReturnType<typeof createIsbot>;
		});
		afterAll(() => {
			jest.restoreAllMocks();
		});
		test("fallback regex detects commong crawlers", () => {
			USER_AGENT_COMMON.forEach((ua) => {
				if (!isbotInstance(ua)) {
					throw new Error(`Failed to detect ${ua} as bot`);
				}
			});
		});
		test("fallback detects gotchas as bots", () => {
			USER_AGENT_GOTCHAS.forEach((ua) => {
				if (!isbotInstance(ua)) {
					throw new Error(`Failed to detect ${ua} as bot (gotcha)`);
				}
			});
		});
		test("fallback does not detect browser as bot", () => {
			expect(isbotInstance(BROWSER_USER_AGENT_EXAMPLE)).toBe(false);
		});
	});

	describe("fixtures", () => {
		test(`✔︎ ${crawlers.length} user agent string should be recognised as crawler`, () => {
			let successCount = 0;
			let misidentifiedStrings: string[] = [];
			crawlers.forEach((crawler) => {
				if (isbot(crawler)) {
					successCount++;
				} else {
					misidentifiedStrings.push(crawler);
				}
			});
			expect(misidentifiedStrings).toEqual([]);
			expect(successCount).toBe(crawlers.length);
		});
		test(`✘ ${browsers.length} user agent string should not be recognised as crawler`, () => {
			let successCount = 0;
			let misidentifiedStrings: string[] = [];
			browsers.forEach((browser) => {
				if (isbot(browser)) {
					misidentifiedStrings.push(browser);
				} else {
					successCount++;
				}
			});
			expect(misidentifiedStrings).toEqual([]);
			expect(successCount).toBe(browsers.length);
		});
	});

	describe("module interface", () => {
		test("interface is as expected", async () => {
			const types = Object.entries(await import("../../src/index")).map(
				([key, value]) => [key, value.constructor.name] as [string, string],
			);
			expect(types).toMatchSnapshot();
		});
		test("regular expressions exports are as expected", () => {
			expect(new RegExp(fullPattern, "i").toString()).toBe(
				getPattern().toString(),
			);
		});
	});
});
