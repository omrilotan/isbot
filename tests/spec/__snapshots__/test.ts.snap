// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isbot module interface interface is as expected 1`] = `
[
  [
    "getPattern",
    "Function",
  ],
  [
    "list",
    "Array",
  ],
  [
    "isbotNaive",
    "Function",
  ],
  [
    "isbot",
    "Function",
  ],
  [
    "createIsbot",
    "Function",
  ],
  [
    "createIsbotFromList",
    "Function",
  ],
  [
    "isbotMatch",
    "Function",
  ],
  [
    "isbotMatches",
    "Function",
  ],
  [
    "isbotPattern",
    "Function",
  ],
  [
    "isbotPatterns",
    "Function",
  ],
]
`;
